### ISSUE TITLE
Remaining duration of -7762 insufficient to load cluster

### ISSUE THREAD

**Issue by edtguy** [awslabs/aws-lambda-redshift-loader#113]
Hi Ian,
I've followed the article to configure and deploy the aws-lambda-redshift-loader, but I'm not sure if I've got it configured correctly. I receive a json file into S3 that contains an array of objects, where the keys of each object correspond exactly with the column names of the Redshift table, with the exception that the json object omits the 'id' column of the table, which is defined as (BIGINT IDENTITY).
Each time a file is put into S3, the process runs, but fails to load Redshift. The first indication of a problem is the line, 'Remaining duration of {n} insufficient to load cluster'. I'm not sure what generates this line, or if it is the root cause, but I've listed the Cloudwatch entries for a sample request below (identity information replaced with {param} placeholders).
Based on the log entries, is the issue something obvious, or would you need additional information?
Thanks,
Ed

`START RequestId: b8af50fa-0045-11e7-b70f-7f92ded88ef5 Version: $LATEST
2017-03-03T19:15:11.136Z	b8af50fa-0045-11e7-b70f-7f92ded88ef5	{"Records": [...}
2017-03-03T19:15:11.381Z	b8af50fa-0045-11e7-b70f-7f92ded88ef5	{"bucket": "...}
2017-03-03T19:15:11.381Z	b8af50fa-0045-11e7-b70f-7f92ded88ef5	Found Redshift Load Configuration for {bucket/prefix}
2017-03-03T19:15:11.396Z	b8af50fa-0045-11e7-b70f-7f92ded88ef5	No open pending Batch 3e952baa-e71b-4e2f-b97d-a3dcddbb45ff
2017-03-03T19:15:11.420Z	b8af50fa-0045-11e7-b70f-7f92ded88ef5	Adding Pending Batch Entry for {bucket/prefix/filename.json}
2017-03-03T19:15:11.529Z	b8af50fa-0045-11e7-b70f-7f92ded88ef5	Current batch count of 1 below batch limit of 2
2017-03-03T19:15:11.529Z	b8af50fa-0045-11e7-b70f-7f92ded88ef5	Current batch age of 0.5290000438690186 seconds below batch timeout of 120
2017-03-03T19:15:11.529Z	b8af50fa-0045-11e7-b70f-7f92ded88ef5	Batch size 68 bytes reached
2017-03-03T19:15:11.646Z	b8af50fa-0045-11e7-b70f-7f92ded88ef5	Creating Manifest for Batch 781532c7-c9e7-4807-939d-917194e68cad
2017-03-03T19:15:11.646Z	b8af50fa-0045-11e7-b70f-7f92ded88ef5	Writing manifest to Manifests/myManifest/manifest-2017-03-03-19:15:11-9951
2017-03-03T19:15:11.861Z	b8af50fa-0045-11e7-b70f-7f92ded88ef5	Created Manifest Manifests/myManifest/manifest-2017-03-03-19:15:11-9951 Successfully
2017-03-03T19:15:11.862Z	b8af50fa-0045-11e7-b70f-7f92ded88ef5	Loading 1 Clusters
2017-03-03T19:15:11.862Z	b8af50fa-0045-11e7-b70f-7f92ded88ef5	Remaining duration of -7762 insufficient to load cluster
2017-03-03T19:15:11.899Z	b8af50fa-0045-11e7-b70f-7f92ded88ef5	Loading Cluster {clusterName}.us-east-1.redshift.amazonaws.com with Lambda credentials
2017-03-03T19:15:11.998Z	b8af50fa-0045-11e7-b70f-7f92ded88ef5	Linked Manifest manifest-2017-03-03-19:15:11-9951 to Batch 781532c7-c9e7-4807-939d-917194e68cad
2017-03-03T19:15:12.067Z	b8af50fa-0045-11e7-b70f-7f92ded88ef5	begin;
COPY {Redshift table} from 's3://Manifests/vivaManifest/manifest-2017-03-03-19:15:11-9951' with credentials as 'aws_access_key_id={accessKey};aws_secret_access_key={secretKey};token={base64Token} manifest format JSON 'auto';
commit;
2017-03-03T19:15:12.067Z	b8af50fa-0045-11e7-b70f-7f92ded88ef5	Connecting to Database {clusterName}.us-east-1.redshift.amazonaws.com:5439
2017-03-03T19:15:12.370Z	b8af50fa-0045-11e7-b70f-7f92ded88ef5	Cluster Load Failure error: Load into table {Redshift table} failed. Check 'stl_load_errors' system table for details. on Cluster {clusterName}.us-east-1.redshift.amazonaws.com
2017-03-03T19:15:12.392Z	b8af50fa-0045-11e7-b70f-7f92ded88ef5	{ '{clusterName}.us-east-1.redshift.amazonaws.com': 
{ status: -1,
error: 
{ [error: Load into table {Redshift table} failed. Check 'stl_load_errors' system table for details.]
name: 'error',
length: 188,
severity: 'ERROR',
code: 'XX000',
detail: undefined,
hint: undefined,
position: undefined,
internalPosition: undefined,
internalQuery: undefined,
where: undefined,
schema: undefined,
table: undefined,
column: undefined,
dataType: undefined,
constraint: undefined,
file: '/home/ec2-user/padb/src/pg/src/backend/commands/commands_copy.c',
line: '2358',
routine: 'DoCopy' } } }
2017-03-03T19:15:12.478Z	b8af50fa-0045-11e7-b70f-7f92ded88ef5	Created new Failed Manifest Manifests/myFail/manifest-2017-03-03-19:15:11-9951
2017-03-03T19:15:12.557Z	b8af50fa-0045-11e7-b70f-7f92ded88ef5
{
    "{clusterName}.us-east-1.redshift.amazonaws.com": {
        "status": -1,
        "error": {
            "name": "error",
            "length": 188,
            "severity": "ERROR",
            "code": "XX000",
            "file": "/home/ec2-user/padb/src/pg/src/backend/commands/commands_copy.c",
            "line": "2358",
            "routine": "DoCopy"
        }
    }
}
2017-03-03T19:15:12.557Z	b8af50fa-0045-11e7-b70f-7f92ded88ef5
{
    "errorMessage": "error"
}
END RequestId: b8af50fa-0045-11e7-b70f-7f92ded88ef5
REPORT RequestId: b8af50fa-0045-11e7-b70f-7f92ded88ef5	Duration: 1457.24 ms	Billed Duration: 1500 ms Memory Size: 128 MB	Max Memory Used: 32 MB	
`

**Comment by alexhogberg** [comment]
It seems to me that the Redshift Loader successfully parses the file and loads it into redshift, but the corresponding table in redshift does not match the input given. I would recommend you to look in the stl_load_errors and see if you can find the corresponding load query and see if you can find the exact load error.

**Comment by edtguy** [comment]
Thanks Alex. I didn't realize I could click on an entry under the load tab and get additional information. The 'error' column under 'Loaded Files' reports, "Invalid JSONPath format: Member is not an object." However, under the 'Raw Line' column, the file contents look correct to me, i.e. an array of one or more objects.
Is this loader expecting a single object per file, rather than an array of objects? Is that why the load files have aborted? If so, that seems a bit surprising, but I could re-write the service that retrieves the json data and writes it to S3 to instead iterate through the array of returned objects, writing one per file.

**Comment by edtguy** [comment]
So I tried writing a single-object json file to S3. It was picked up by the Lambda process, but there is no indication of activity in the Redshift load listing. CloudWatch reported the following:
`START RequestId: a03026a4-02ac-11e7-b024-f3814a03db71 Version: $LATEST
2017-03-06T20:36:50.444Z	a03026a4-02ac-11e7-b024-f3814a03db71	{    "Records": [...]}
2017-03-06T20:36:51.320Z	a03026a4-02ac-11e7-b024-f3814a03db71	{"bucket": ... }
2017-03-06T20:36:51.320Z	a03026a4-02ac-11e7-b024-f3814a03db71	Found Redshift Load Configuration for {bucket/prefix}
2017-03-06T20:36:51.572Z	a03026a4-02ac-11e7-b024-f3814a03db71	Adding Pending Batch Entry for {bucket/prefix/file.json}
2017-03-06T20:36:51.781Z	a03026a4-02ac-11e7-b024-f3814a03db71	Reload of Configuration Complete after attempting to write to Locked Batch 4a4b63f3-bb03-4200-92b4-4dbff8bdd795. Attempt 1
2017-03-06T20:36:51.944Z	a03026a4-02ac-11e7-b024-f3814a03db71	Reload of Configuration Complete after attempting to write to Locked Batch 4a4b63f3-bb03-4200-92b4-4dbff8bdd795. Attempt 1
2017-03-06T20:36:52.119Z	a03026a4-02ac-11e7-b024-f3814a03db71	Reload of Configuration Complete after attempting to write to Locked Batch 4a4b63f3-bb03-4200-92b4-4dbff8bdd795. Attempt 1
2017-03-06T20:36:52.278Z	a03026a4-02ac-11e7-b024-f3814a03db71	Reload of Configuration Complete after attempting to write to Locked Batch 4a4b63f3-bb03-4200-92b4-4dbff8bdd795. Attempt 1
2017-03-06T20:36:52.397Z	a03026a4-02ac-11e7-b024-f3814a03db71	Reload of Configuration Complete after attempting to write to Locked Batch 4a4b63f3-bb03-4200-92b4-4dbff8bdd795. Attempt 1
2017-03-06T20:36:52.518Z	a03026a4-02ac-11e7-b024-f3814a03db71	Reload of Configuration Complete after attempting to write to Locked Batch 4a4b63f3-bb03-4200-92b4-4dbff8bdd795. Attempt 1
2017-03-06T20:36:52.639Z	a03026a4-02ac-11e7-b024-f3814a03db71	Reload of Configuration Complete after attempting to write to Locked Batch 4a4b63f3-bb03-4200-92b4-4dbff8bdd795. Attempt 1
2017-03-06T20:36:52.740Z	a03026a4-02ac-11e7-b024-f3814a03db71	Reload of Configuration Complete after attempting to write to Locked Batch 4a4b63f3-bb03-4200-92b4-4dbff8bdd795. Attempt 1
2017-03-06T20:36:52.810Z	a03026a4-02ac-11e7-b024-f3814a03db71	Reload of Configuration Complete after attempting to write to Locked Batch 4a4b63f3-bb03-4200-92b4-4dbff8bdd795. Attempt 1
2017-03-06T20:36:52.898Z	a03026a4-02ac-11e7-b024-f3814a03db71	Reload of Configuration Complete after attempting to write to Locked Batch 4a4b63f3-bb03-4200-92b4-4dbff8bdd795. Attempt 1
2017-03-06T20:36:52.983Z	a03026a4-02ac-11e7-b024-f3814a03db71	Reload of Configuration Complete after attempting to write to Locked Batch 4a4b63f3-bb03-4200-92b4-4dbff8bdd795. Attempt 1
2017-03-06T20:36:53.080Z	a03026a4-02ac-11e7-b024-f3814a03db71	Reload of Configuration Complete after attempting to write to Locked Batch 4a4b63f3-bb03-4200-92b4-4dbff8bdd795. Attempt 1
2017-03-06T20:36:53.173Z	a03026a4-02ac-11e7-b024-f3814a03db71	Reload of Configuration Complete after attempting to write to Locked Batch 4a4b63f3-bb03-4200-92b4-4dbff8bdd795. Attempt 1
2017-03-06T20:36:53.244Z	a03026a4-02ac-11e7-b024-f3814a03db71	Reload of Configuration Complete after attempting to write to Locked Batch 4a4b63f3-bb03-4200-92b4-4dbff8bdd795. Attempt 1
2017-03-06T20:36:53.415Z	a03026a4-02ac-11e7-b024-f3814a03db71	Reload of Configuration Complete after attempting to write to Locked Batch 4a4b63f3-bb03-4200-92b4-4dbff8bdd795. Attempt 1
END RequestId: a03026a4-02ac-11e7-b024-f3814a03db71
REPORT RequestId: a03026a4-02ac-11e7-b024-f3814a03db71	Duration: 3002.52 ms	Billed Duration: 3000 ms Memory Size: 128 MB	Max Memory Used: 19 MB	
2017-03-06T20:36:53.445Z a03026a4-02ac-11e7-b024-f3814a03db71 Task timed out after 3.00 seconds
`
I'm not sure what to make of this.

**Comment by alexhogberg** [comment]
As per your first comment: I don't have any experience using JSON since we use CSV to load files but there we are able to use multiple lines per file and load them successfully. Perhaps @IanMeyers can explain how json input should look. You could perhaps look at [this reference](http://docs.aws.amazon.com/redshift/latest/dg/copy-usage_notes-copy-from-json.html) and see if you can alter your input when loading data into redshift.

As per your second command: It tried to write the file into a batch with status locked and failed to complete the task. You need to manually try to unlock the batch using what is described [here](https://github.com/awslabs/aws-lambda-redshift-loader/#unlocking-a-batch).

**Comment by edtguy** [comment]
Thanks Alex! I  was able to use the utilities to unlock the locked batches, and the documentation for JSON showed that the loader doesn't understand an array of JSON objects (i.e. a collection), so I had to add in this hack before sending the data to S3 via the putObject call:
`var RedshiftHack = _.map(JSON.parse(data),function(f) {return JSON.stringify(f);}).join('\n');`
The above essentially strips the enclosing brackets off the array and replaces the commas between objects with line breaks.
Lambda is now successfully loading the JSON data into Redshift. Thanks again!

**Comment by charlesfeng** [comment]
Sorry for reviving a dead thread. For anyone else who is seeing this issue, ours was caused by the Lambda function timeout being set too low (3 seconds), and so this line https://github.com/awslabs/aws-lambda-redshift-loader/blob/e979e0533bc51c4c5c1ddbc41d9f06313d2ebc15/index.js#L1012 which subtracts 10 sec from the remaining timeout would constantly trigger batch timeouts. We updated our Lambda timeout to 1 minute and the issue was resolved.
